json/json.go

```diff
@@ -115,9 +116,21 @@ func (c *decoder) Decode(v interface{}) error {
 		if err != nil {
 			return err
 		}
+		return json.NewDecoder(r).Decode(v)
 	}
 
-	return json.NewDecoder(r).Decode(v)
+	decoder := json.NewDecoder(r)
+	err := decoder.Decode(v)
+	// the JSON decoder may read ahead so reset the reader so we don't
+	// drop the extra bytes read
+	c.r = io.MultiReader(decoder.Buffered(), c.r)
```

Comment:
This will end up creating a large chain of readers that'll build up over time. Instead, I think we need to store the decoder itself and construct a new `io.Multireader(decoder.Buffered(), c.r)` on the fly (unless you can think of a better solution...).